### PR TITLE
Fix generating base keys from empty indexing paging state

### LINKED ISSUE #6134: secondary_indexes_test: test_truncate_base: Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing"
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/447/testReport/junit/secondary_indexes_test/TestSecondaryIndexes/test_truncate_base/
Scylla version a15b32c9d9668928aaddd7b8491c63322b1aa578
```
Error Message
('Unable to complete the operation against any hosts', {<Host: 127.0.29.2:9042 dc1>: <Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing">, <Host: 127.0.29.3:9042 dc1>: <Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing">, <Host: 127.0.29.4:9042 dc1>: <Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing">, <Host: 127.0.29.1:9042 dc1>: <Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing">})
Stacktrace
Traceback (most recent call last):
  File "/usr/lib64/python3.7/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib64/python3.7/unittest/case.py", line 645, in run
    testMethod()
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/secondary_indexes_test.py", line 759, in test_truncate_base
    cl=ConsistencyLevel.QUORUM)
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/dtest.py", line 1311, in inner
    return func(*args, **kwargs)
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/assertions.py", line 86, in assert_all
    list_res = _get_list_res(session, query, cl, ignore_order, result_as_string)
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/assertions.py", line 51, in _get_list_res
    res = session.execute(simple_query)
  File "cassandra/cluster.py", line 2611, in cassandra.cluster.Session.execute
  File "cassandra/cluster.py", line 4829, in cassandra.cluster.ResponseFuture.result
cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.0.29.2:9042 dc1>: <Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing">, <Host: 127.0.29.3:9042 dc1>: <Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing">, <Host: 127.0.29.4:9042 dc1>: <Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing">, <Host: 127.0.29.1:9042 dc1>: <Error from server: code=0000 [Server error] message="Column key was expected to be provided in the index clustering key, but the whole index clustering key is missing">})
```

This exception was added in b3fdb742aed5ef1b8d849ef099a1e2044561831e for #5856

### PR DESCRIPTION
**Pull request by psarna**
An empty partition/clustering key pair is a valid state of the 
query paging state. Unfortunately, recent attempts at debugging
a flaky test (#5856) resulted in introducing an assertion (7616290)
which breaks when trying to generate a key from such a pair.
In order to keep the assertion (since it still makes sense in its 
scope), but at the same time translate empty keys properly,
empty keys are now explicitly processed at the beginning of the 
function.
This behaviour was 100% reproducible in a secondary index dtest below.

Fixes #6134
Refs #5856
Tests: unit(dev),
     dtest(TestSecondaryIndexes.test_truncate_base)

### CODE REVIEW

File: cql3/statements/select_statement.cc [hunk 1]
```diff
@@ -514,25 +514,26 @@ indexed_table_select_statement::do_execute_base_query(
...
-                    assert(old_paging_state->get_clustering_key().has_value());
-                    auto base_ck = generate_base_key_from_index_pk<clustering_key>(old_paging_state->get_partition_key(),
-                            old_paging_state->get_clustering_key(), *_schema, *_view_schema);
-                    command->slice.set_range(*_schema, base_pk,
-                            std::vector<query::clustering_range>{query::clustering_range::make_starting_with(range_bound<clustering_key>(base_ck, false))});
-                } else {
-                    command->slice.set_range(*_schema, base_pk, std::vector<query::clustering_range>{query::clustering_range::make_open_ended_both_sides()});
-                }
-            }
+            auto command = ::make_lw_shared<query::read_command>(*base_command);
```
Comment (nyh) [inline cql3/statements/select_statement.cc:541]: Something looks wrong to me here - or maybe I just don't understand here.
In the old code, for the first iteration (concurrency==1), the code copied "cmd" and modified it based on the old paging state. But when concurrency > 1, the code also copied "cmd", but didn't modify it.
In the new code, all iterations copy the same base_command. What am I missing?
Also, do we have any test that can verify that this code actually works as intended?

Comment (psarna) [reply cql3/statements/select_statement.cc:541]: Oh snap, you're right... that's what I get for pushing patches too late in the evening. Only the first command should be trimmed, the rest should be a copy of the original command and this patch does not make sense. My original (correct) intention was that if we have an old paging state and our query spans multiple partitions, the first subquery may only need to fetch a small slice of the partition, depending of where we ended last time, but all the next queries should use the original, full range. I remember that there are dtests for that, I'll try to find out which ones.

**Comment by psarna** [comment]
v2 changes:
 * removed the buggy "extraction out of the loop" commit
 * added regression unit test cases - which are verified to fail before the patch and succeed after the patch

**Comment by nyh** [comment]
Committed to next.

**Comment by eliransin** [comment]
@nyh @psarna , AFAICU there is no risk to backport it to 4.0 right? But should we also backport 
b3fdb742aed5ef1b8d849ef099a1e2044561831e before that? or should we avoid both of them altogether?

**Comment by psarna** [comment]
Avoiding both is also ok. It will mean that we'll lose some debuggability *if* there's another problem with generating base keys, but it's not critical to backport.

**Pull request merged**